torchvision/ops/poolers.py

```diff
@@ -106,6 +107,135 @@ def _infer_scale(feature: Tensor, original_size: List[int]) -> float:
     return possible_scales[0]
 
 
+def _setup_scales(
+    features: List[Tensor], image_shapes: List[Tuple[int, int]], canonical_scale: int, canonical_level: int
+) -> Tuple[List[float], LevelMapper]:
+    assert len(image_shapes) != 0
+    max_x = 0
+    max_y = 0
+    for shape in image_shapes:
+        max_x = max(shape[0], max_x)
+        max_y = max(shape[1], max_y)
+    original_input_shape = (max_x, max_y)
+
+    scales = [_infer_scale(feat, original_input_shape) for feat in features]
+    # get the levels in the feature map by leveraging the fact that the network always
+    # downsamples by a factor of 2 at each level.
+    lvl_min = -torch.log2(torch.tensor(scales[0], dtype=torch.float32)).item()
+    lvl_max = -torch.log2(torch.tensor(scales[-1], dtype=torch.float32)).item()
+
+    map_levels = initLevelMapper(
+        int(lvl_min),
+        int(lvl_max),
+        canonical_scale=canonical_scale,
+        canonical_level=canonical_level,
+    )
+    return scales, map_levels
+
+
+@torch.fx.wrap
+def _multiscale_roi_align(
+    x: Dict[str, Tensor],
+    boxes: List[Tensor],
+    image_shapes: List[Tuple[int, int]],
+    featmap_names: List[str],
+    canonical_scale: int,
+    canonical_level: int,
+    output_size: List[int],
+    sampling_ratio: int,
+    scales: Optional[List[float]] = None,
+) -> Tensor:
+    """
+    Args:
+        x (OrderedDict[Tensor]): feature maps for each level. They are assumed to have
+            all the same number of channels, but they can have different sizes.
+        boxes (List[Tensor[N, 4]]): boxes to be used to perform the pooling operation, in
+            (x1, y1, x2, y2) format and in the image reference size, not the feature map
+            reference. The coordinate must satisfy ``0 <= x1 < x2`` and ``0 <= y1 < y2``.
+        image_shapes (List[Tuple[height, width]]): the sizes of each image before they
+            have been fed to a CNN to obtain feature maps. This allows us to infer the
+            scale factor for each one of the levels to be pooled.
+        canonical_scale (int): canonical_scale for LevelMapper
+        canonical_level (int): canonical_level for LevelMapper
+        output_size (Union[List[Tuple[int, int]], List[int]]): size of the output
+        sampling_ratio (int): sampling ratio for ROIAlign
+        scales (Optional[List[float]]): If None, scales will be automatically infered. Default value is None.
+    Returns:
+        result (Tensor)
+    """
+
+    x_filtered = []
+    for k, v in x.items():
+        if k in featmap_names:
+            x_filtered.append(v)
+    num_levels = len(x_filtered)
+    rois = _convert_to_roi_format(boxes)
+
+    if scales is None:
+        scales, mapper = _setup_scales(x_filtered, image_shapes, canonical_scale, canonical_level)
```

Comment:
Neither `scales` nor `mapper` are passed by reference, so as far as I understand you end up calling `_setup_scales()` every time. I believe this needs to be refactored to return every value we end up modifying and assigning them to the original `self` parameters.

Reply:
You are right they are set to None every time in the `MultiScaleRoIAlign` constructor so this would be always called. But in that case I suggest to change this to 
```python
scales, mapper = _setup_scales(x_filtered, image_shapes, canonical_scale, canonical_level)
```
instead of 
```python
if scales is None:
        scales, mapper = _setup_scales(x_filtered, image_shapes, canonical_scale, canonical_level)
else:
        mapper = None
```

and also remove `scales` from the arguments list.

But it is not clear to me why we would need to return the values and assigning them to the original `self` paramters.

Reply:
Please ignore this, because of https://github.com/pytorch/vision/pull/4810/files#r754140003 this comment no longer applies.